Multi Step Form 과제 2

pereng11 opened the pull request:
## 09.27 리뷰 반영 수정 사항

### step&schema 상태관리 재설계

#### 변경 전
- step과 각 step의 schema를 결합하여 funnel state로 관리(useFunnel)
- 렌더링 분기처리 컴포넌트가 useFunnel에 종속된 형태
- 모든 step의 schema는 이전 step의 필드를 포함하는 것으로 타입 안정성 보장
- 임시저장 로직 구현
  - step은 queryparam에 저장
  - schema는 로컬스토리지에 저장

#### 변경 후
- step / schema를 관리하는 각각의 훅과 그 둘을 통합하여 Review의 전반적인 동작을 관리하는 훅으로 레이어 분리
  - level1: 도메인에 종속되지 않는 훅
    - useStep
  - level2: 특정 도메인에 종속되어 한가지 관심사만 다루는 훅. 필요한 경우 level1 훅 사용
    - useReviewStep / useReviewSchema 
  - level3: level2 훅들을 조합하여 페이지의 전반적인 동작을 제어하는 훅
    - useReviewFunnel
- 렌더링 분기처리만 담당하는 공통 컴포넌트 `SwitchCase` 구현
- 모든 스텝을 통합한 폼 상태에 초기값을 부여하는 것으로 타입 안정성 보장 + 각 스텝의 schema는 실제로 다루는 값만 가지도록 수정

---

## 09.02 리뷰 반영 수정 사항

- 스텝 컴포넌트 내 하드코딩 된 타 스텝 의존성 제거
- nullish한 값 명시적으로 확인하도록 수정
- 값 검증 로직 UI컴포넌트 -> zod schema로 이동

---

## 리뷰 반영한 설계 내용 & 고민

- form 관련 역할 분리
  - 타입 & 유효성 검사는 types 디렉토리의 zod 스키마를 생성하는 곳에 일임한다.
  - view 의 컴포넌트에서는 form 요소의 동작을 제어하는 것에 집중한다.
  - 리뷰해주신 내용 기반으로 고민해본 결과, 스키마를 생성하는 곳에서 유효성 검사까지 가져가는 것이 코드의 응집도 측면에서 유리하고, 그 과정에서 zod의 API를 자연스럽게 활용할 수 있다고 판단했습니다.

- funnel 데이터 통합
  - 이번 과제의 요구사항 기준으로는 퍼널 간 직접적인 의존 관계를 지우고 필요한 필드를 누적해서 전달하는 것이 가장 적절하다고 판단했습니다. 데이터 전달 과정을 직관적으로 파악할 수 있고, zod 스키마의 output타입과 유효성 검사로 전달하는 필드의 정합성을 보장할 수 있다고 생각했습니다.
  - 만약 퍼널 개수&다루는 필드가 많아지거나 퍼널 간 이동 경로가 복잡해진다면, 각 퍼널의 완료 여부와 완료 시점의 상태 스냅샷을 지금보다 편리하고 직관적으로 다룰 수 있도록 펴널 모듈을 고도화하고 마지막 퍼널에서 직접 필요한 상태를 꺼내 쓰는 방법도 적용해볼 수 있을 것 같습니다.
  
- 적정 수준의 추상화
  - 퍼널 모듈에서 불필요하게 추상화된 코드를 걷어내보려고 했는데 생각보다 쉽지 않았습니다. :disappointed_relieved: 퍼널 훅의 input/output 인터페이스를 제외하면 history 객체의 퍼널 라우팅 함수들이 남는데, 기능의 완결성 측면에서 push / replace / go / back 중 일부만 제거하는 것도 조금 이상하게 느껴졌습니다. 혹시 어떤 부분에서 추상화가 과도하다고 생각하셨는지 좀 더 자세히 설명 부탁드려도 될까요?


## 구현 내용

- step3-5 구현
- step1-5 유효성 검사 로직 적용&개선
- 책목록 조회 / 리뷰 저장 API 연동
  - tanstack query 사용
  - suspense / error boundary 적용


<img width="369" height="423" alt="image" src="https://github.com/user-attachments/assets/9642534c-f26c-4d75-919f-02e7f6174a7d" />
<img width="454" height="228" alt="image" src="https://github.com/user-attachments/assets/599df29e-d725-4345-bdfb-582127111585" />
<img width="431" height="474" alt="image" src="https://github.com/user-attachments/assets/d1cc20b6-d64f-450d-91f2-406dfee70323" />
<img width="746" height="423" alt="image" src="https://github.com/user-attachments/assets/bde7100c-4ea3-43fa-9f3e-96ac66e5e1ec" />

pereng11 committed:
feat: step3 구현

- 독후감 입력 폼
- rating에 따른 유효성 검사

pereng11 committed:
fix: step1 유효성 검사 로직 수정

- 유효성 검사 케이스 분리
- 독서 상태-기간 사이 검사 타이밍 & 값 어긋나지 않도록 수정
- 날짜 비교 로직 수정
- 공통 에러메시지 컴포넌트 적용

pereng11 committed:
fix: step2 유효성 검사 로직 수정

pereng11 committed:
fix: step3 유효성 검사 로직 수정

pereng11 committed:
feat: step4 - 인용구 입력 단계 구현

pereng11 committed:
feat: step5 - 공개여부 입력 단계 폼 구현

pereng11 committed:
feat: Review 관련 타입 위계 조정

- 서버에서 가져오는 타입 api 계층으로 이동

pereng11 committed:
feat: Book 관련 타입 위계 조정

- 서버에서 가져오는 타입 api 계층으로 이동

pereng11 committed:
feat: Review 저장 로직 구현

- DB연동 없이 API만 연결

pereng11 committed:
feat: 책 목록 페이지 구현

pereng11 committed:
feat: 책 목록 - 리뷰 작성 페이지 연결

pereng11 committed:
refactor: 컴포넌트 폴더 구조 변경

pereng11 committed:
feat: Suspense & ErrorBoundary를 통합한 바운더리 컴포넌트 구현 및 적용

pereng11 committed:
refactor: 오타 수정

brightparagon commented on src/components/new/view/ReviewFunnel.tsx:
```diff
@@ -0,0 +1,40 @@
+
+  const book = getBookById(bookId ?? "");
+
+  if (!book) {
```
null 및 value check는 js 문법적 특성을 이용한 축약된 표현보다는 분명한 로직으로 표현하는 것이 좋습니다!

brightparagon commented on src/components/new/view/ReviewFunnel.tsx:
```diff
@@ -0,0 +1,40 @@
+
+  return (
+    <div>
+      <Render
+        step1={(props) => <BasicStep {...props} book={book} />}
+        step2={RecommendStep}
+        step3={BookReportStep}
+        step4={(props) => <QuoteStep {...props} book={book} />}
+        step5={PublishStep}
+      />
+    </div>
+  );
```
저는 Render를 훅에서 제공하지 않고, 독립적으로 아래 같은 방식으로도 만들 수 있겠다 싶었습니다 ㅎㅎ 그러면 step은 훅 혹은 params에서 가져오고, step을 관리하는 훅이 있고 그걸 독립적으로 받아서 스텝에 맞는 컴포넌트를 렌더링하는 구조

```typescript
<SwitchCases
  value={step}
  cases={{
    step1: <Step1 />
    ...
  }}
/>
```

pereng11 replied:
그것도 좋은 방법인 것 같습니다!
예시로 적어주신 SwitchCases는 multi-step-form 형태가 아닌 다른 렌더링 분기처리에도 범용적으로 쓸 수 있는 장점이 있겠네요👍

다만 해당 컴포넌트로 분기처리를 할 때는 form의 진행 상태에 따라 undefined일 수 있는 값에 대한 타입 보장이 어려울 것 같습니다
(예를 들면 step2의 평점을 step3에서 참조해야 하는 경우 등).
이런 저런 시도를 해봤는데 깔끔하게 처리할만한 방법을 못찾았네요🥲 혹시 생각하신 방안이 있다면 설명 부탁드려도 괜찮을까요?

brightparagon replied:
폼 마다 defaultValues를 준다면 타입 보장 및 초기값에 대한 문제에서는 벗어날 수 있지 않을까요?

brightparagon replied:
스텝 간 의존성이 있는 경우도 해결할 수 있습니다

brightparagon commented:
👍

pereng11 committed:
refactor: 값 검증 로직 zod로 위임

pereng11 committed:
refactor: nullish 명시적으로 확인하도록 수정

pereng11 committed:
fix: package.json 오타 수정

pereng11 committed:
fix: QuoteSchema 값 전처리 로직 수정

pereng11 committed:
refactor: 스텝 컴포넌트 내부 하드코딩된 의존성 제거

- 스텝 컴포넌트에서는 정해진 타입에 맞는 데이터만 반환하도록 수정
- 스텝 컴포넌트를 사용하는 곳에서 일괄적으로 흐름 관리

pereng11 committed:
feat: 공통 컴포넌트 구현 - SwitchCase

pereng11 committed:
feat: 공통 hook 구현 - useStep

- 스텝 앞/뒤 이동 공통 로직

pereng11 committed:
refactor: 리뷰 상태관리 컨셉 수정

- 변경 전
  - step과 각 step의 chema를 결합하여 funnel state로 관리
- 변경 후
  - step / schema를 관리하는 각각의 훅과 그 둘을 통합하여 Review의 전반적인 동작을 관리하는 훅으로 레이어 분리

pereng11 committed:
feat: [홈] 책 선택 영역 확대

pereng11 committed:
feat: 스텝 단위 임시저장 기능 구현

- 로컬스토리지 사용
- 스텝 완료 시점에 임시 저장

pereng11 committed:
fix: 별점 초기값 올바르게 적용되도록 수정
